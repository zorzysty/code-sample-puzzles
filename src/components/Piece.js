import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { DragSource } from 'react-dnd';
import imgSrc from '../assets/picture.jpg';
import { DIMENSIONS } from '../const';


const spec = {
  beginDrag: (props) => {
    props.handleDrag();
    return props.piece;
  },
  endDrag: (props, monitor) => {
    // return piece to original position when dropped outside of target
    if (!monitor.didDrop()) return;

    // pass piece to handleDrop method if placed correctly and false otherwise
    props.handleDrop(monitor.getDropResult().success && props.piece);
  },
};

function collect(connect, monitor) {
  return {
    connectDragSource: connect.dragSource(),
    isDragging: monitor.isDragging(),
  };
}

class Piece extends PureComponent {
  render() {
    const {
      isDragging, // eslint-disable-line react/prop-types
      connectDragSource, // eslint-disable-line react/prop-types
      style,
      piece,
    } = this.props;

    // hide piece in drawer while it's being dragged
    const opacity = isDragging ? 0 : 1;
    // prevent overlaying of images while dragging
    const zIndex = isDragging ? 0 : 10;
    // set proper piece image using CSS sprites
    const left = -(piece.id % DIMENSIONS.columns) * DIMENSIONS.pieceSize;
    const top = -Math.floor(piece.id / DIMENSIONS.rows) * DIMENSIONS.pieceSize;

    return connectDragSource(
      <div
        className="Piece"
        style={{
          background: `url(${imgSrc})`,
          backgroundPosition: `${left}px ${top}px`,
          backgroundSize: `${DIMENSIONS.columns * 100}%`,
          opacity,
          zIndex,
          ...style,
        }}
      />,
    );
  }
}

export default DragSource('piece', spec, collect)(Piece);

Piece.propTypes = {
  piece: PropTypes.shape({
    id: PropTypes.number.isRequired,
  }).isRequired,
  handleDrop: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  handleDrag: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
  style: PropTypes.objectOf(PropTypes.string),
};

Piece.defaultProps = {
  handleDrop: () => {},
  handleDrag: () => {},
  style: {},
};
