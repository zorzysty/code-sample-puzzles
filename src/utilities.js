import { DIMENSIONS } from './const';

/**
 * shuffleArray
 *
 * Shuffles item from provided array and returns them as new array without mutating original one
 *
 * @param inputArray {Array} Array to be shuffled
 * @returns {Array}
 */
export function shuffleArray(inputArray) {
  const array = [...inputArray];
  for (let i = array.length - 1; i > 0; i -= 1) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}

/**
 * getRandomInt
 *
 * Returns pseudo-random integer from provided range
 *
 * @param min {Number} minimal expected value
 * @param max {Number} maximal expected value
 * @returns {Number}
 */
export function getRandomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

/**
 * getRandomPositions
 *
 * Generates array of objects containing top and left properties with generated pseudo-random values
 *
 * @param length {Number}
 * @returns {Array}
 */
export function getRandomPositions(length) {
  const { pieceSize, drawerPadding: padding, width } = DIMENSIONS;

  return [...Array(length)].map(() => ({
    top: `${getRandomInt(padding, pieceSize / 2 - padding)}px`,
    left: `${getRandomInt(padding, width - pieceSize - padding)}px`,
  }));
}
