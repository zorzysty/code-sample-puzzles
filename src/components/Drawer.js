import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Piece from './Piece';
import { getRandomPositions } from '../utilities';


class Drawer extends PureComponent {
  constructor(props) {
    super(props);

    // generate random positions to put pieces in within drawer
    this.positions = getRandomPositions(props.pieces.length);
  }

  render() {
    const {
      handleDrop, pieces, handleDrag,
    } = this.props;

    return pieces.length > 0 && (
      <div className="Drawer">
        {pieces.map(piece => (
          <Piece
            key={piece.id}
            piece={piece}
            handleDrop={handleDrop}
            handleDrag={handleDrag}
            style={this.positions[piece.id]}
          />
        ))}
      </div>
    );
  }
}

Drawer.propTypes = {
  pieces: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
    }).isRequired,
  ).isRequired,
  handleDrop: PropTypes.func.isRequired,
  handleDrag: PropTypes.func.isRequired,
};

export default Drawer;
