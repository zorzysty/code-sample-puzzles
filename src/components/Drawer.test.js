import React from 'react';
import { shallow } from 'enzyme';
import Drawer from './Drawer';
import { getRandomPositions } from '../utilities';
import Piece from './Piece';


const values = [
  {
    top: '40px',
    left: '253px',
  },
  {
    top: '53px',
    left: '182px',
  },
  {
    top: '45px',
    left: '305px',
  },
  {
    top: '31px',
    left: '222px',
  },
  {
    top: '53px',
    left: '323px',
  },
  {
    top: '64px',
    left: '269px',
  },
  {
    top: '73px',
    left: '79px',
  },
  {
    top: '48px',
    left: '266px',
  },
];
const pieces = [
  { id: 0 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }, { id: 6 }, { id: 7 }, { id: 8 },
];
const handleDragMock = jest.fn();
const handleDropMock = jest.fn();

jest.mock('../utilities');
getRandomPositions.mockReturnValue(values);

describe('Drawer', () => {
  const component = shallow(
    <Drawer
      pieces={pieces}
      handleDrag={handleDragMock}
      handleDrop={handleDropMock}
    />,
  );

  it('should render correctly with pieces', () => {
    expect(component)
      .toMatchSnapshot();
  });

  it('should contain proper Piece components', () => {
    expect(component.containsAllMatchingElements(pieces.map(piece => <Piece piece={piece} />)))
      .toBe(true);

    expect(component.containsMatchingElement(<Piece piece={{ id: 1 }} />))
      .toBe(false);
  });

  it('should render correctly without pieces', () => {
    const emptyComponent = shallow(
      <Drawer pieces={[]} handleDrag={() => {}} handleDrop={() => {}} />,
    );

    expect(emptyComponent.containsMatchingElement(<Piece />))
      .toBe(false);

    expect(emptyComponent)
      .toMatchSnapshot();
  });

  it('should be able to call functions passed as props to Piece components', () => {
    const piece = component.find(Piece).first();

    piece.prop('handleDrag')();
    expect(handleDragMock)
      .toHaveBeenCalledTimes(1);

    piece.prop('handleDrop')();
    expect(handleDropMock)
      .toHaveBeenCalledTimes(1);
  });
});
