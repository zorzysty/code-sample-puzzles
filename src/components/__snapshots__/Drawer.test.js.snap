// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer should render correctly with pieces 1`] = `ShallowWrapper {}`;

exports[`Drawer should render correctly without pieces 1`] = `ShallowWrapper {}`;
